pilot/cmd/pilot-agent/main.go

```diff
@@ -125,74 +122,39 @@ var (
-			if err := agent.Start(); err != nil {
-				log.Fatala("Agent start up error", err)
-			}
-
 			// If we are using a custom template file (for control plane proxy, for example), configure this.
 			if templateFile != "" && proxyConfig.CustomConfigFile == "" {
 				proxyConfig.ProxyBootstrapTemplatePath = templateFile
 			}
 
+			envoyOptions := envoy.ProxyConfig{
+				LogLevel:          proxyLogLevel,
+				ComponentLogLevel: proxyComponentLogLevel,
+				LogAsJSON:         loggingOptions.JSONEncoding,
+				NodeIPs:           proxy.IPAddresses,
+				Sidecar:           proxy.Type == model.SidecarProxy,
+				OutlierLogPath:    outlierLogPath,
+			}
+			agentOptions := options.NewAgentOptions(proxy)
+			agent := istio_agent.NewAgent(proxyConfig, agentOptions, secOpts, envoyOptions)
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
 
+			// Start in process SDS, dns server, and xds proxy.
+			if err := agent.Start(ctx); err != nil {
+				log.Fatala("Agent start up error", err)
+			}
+
 			// If a status port was provided, start handling status probes.
 			if proxyConfig.StatusPort > 0 {
 				if err := initStatusServer(ctx, proxy, proxyConfig, agent); err != nil {
 					return err
 				}
 			}
 
-			provCert := agent.FindRootCAForXDS()
-			if provCert == "" {
-				// Envoy only supports load from file. If we want to use system certs, use best guess
-				// To be more correct this could lookup all the "well known" paths but this is extremely \
-				// unlikely to run on a non-debian based machine, and if it is it can be explicitly configured
-				provCert = "/etc/ssl/certs/ca-certificates.crt"
-			}
-			node, err := bootstrap.GetNodeMetaData(bootstrap.MetadataOptions{
-				ID:                  proxy.ServiceNode(),
-				Envs:                os.Environ(),
-				Platform:            platform.Discover(),
-				InstanceIPs:         proxy.IPAddresses,
-				StsPort:             stsPort,
-				ProxyConfig:         proxyConfig,
-				ProxyViaAgent:       agentOptions.ProxyXDSViaAgent,
-				PilotSubjectAltName: pilotSAN,
-				OutlierLogPath:      outlierLogPath,
-				PilotCertProvider:   secOpts.PilotCertProvider,
-				ProvCert:            provCert,
-			})
-			if err != nil {
-				log.Error("Failed to extract node metadata: ", err)
-				os.Exit(1)
-			}
-			envoyProxy := envoy.NewProxy(envoy.ProxyConfig{
-				Node:              node,
-				LogLevel:          proxyLogLevel,
-				ComponentLogLevel: proxyComponentLogLevel,
-				LogAsJSON:         loggingOptions.JSONEncoding,
-				NodeIPs:           proxy.IPAddresses,
-				Sidecar:           proxy.Type == model.SidecarProxy,
-			})
-
-			drainDuration, _ := types.DurationFromProto(proxyConfig.TerminationDrainDuration)
-			envoyAgent := envoy.NewAgent(envoyProxy, drainDuration)
 			// On SIGINT or SIGTERM, cancel the context, triggering a graceful shutdown
-			go cmd.WaitSignalFunc(cancel)
-
-			return envoyAgent.Run(ctx)
+			cmd.WaitSignalFunc(cancel)
```

Comment:
Its a bit concerning no tests fail, but I think this broke the graceful shutdown (maybe not this line exactly, but the PR).

With this PR:
```
^C2021-04-22T07:47:42.469739Z   warning envoy main      caught SIGINT
2021-04-22T07:47:42.472949Z     warning envoy config    StreamAggregatedResources gRPC config stream closed: 13,
2021-04-22T07:47:42.474382Z     warning envoy config    StreamSecrets gRPC config stream closed: 13,
2021-04-22T07:47:42.474458Z     warning envoy config    StreamSecrets gRPC config stream closed: 13,
```

Without this PR:
```
^C2021-04-22T07:48:05.557807Z   warning envoy main      caught SIGINT
2021-04-22T14:48:05.557983Z     info    Agent draining Proxy
2021-04-22T14:48:05.557983Z     info    Status server has successfully terminated
2021-04-22T14:48:05.558239Z     error   accept tcp [::]:15020: use of closed network connection
2021-04-22T07:48:05.560170Z     warning envoy config    StreamAggregatedResources gRPC config stream closed: 13,
2021-04-22T07:48:05.561390Z     warning envoy config    StreamSecrets gRPC config stream closed: 13,
2021-04-22T07:48:05.561468Z     warning envoy config    StreamSecrets gRPC config stream closed: 13,
2021-04-22T14:48:05.562976Z     info    ads     ADS: "@" .-2 terminated rpc error: code = Canceled desc = context canceled
2021-04-22T14:48:05.562982Z     info    ads     ADS: "@" .-1 terminated rpc error: code = Canceled desc = context canceled
2021-04-22T14:48:05.574072Z     info    failed draining listeners for Envoy on port 15000: Post "http://127.0.0.1:15000/drain_listeners?inboundonly&graceful": read tcp 127.0.0.1:43376->127.0.0.1:15000: read: connection reset by peer
2021-04-22T14:48:05.574139Z     warn    Error in invoking drain listeners endpoint Post "http://127.0.0.1:15000/drain_listeners?inboundonly&graceful": read tcp 127.0.0.1:43376->127.0.0.1:15000: read: connection reset by peer
2021-04-22T14:48:05.574161Z     info    Graceful termination period is 3s, starting...
2021-04-22T14:48:08.576413Z     info    Graceful termination period complete, terminating remaining proxies.
2021-04-22T14:48:08.576450Z     warn    Aborted all epochs
2021-04-22T14:48:08.576459Z     info    Agent has successfully terminated
```

This is with `terminationDrainDuration: 3s` configured

Reply:
oh probably since we don't block on agent.Run() completetion

Reply:
Fixed.